design-documents/simplified-module-declaration.md

```diff
@@ -0,0 +1,29 @@
+# Simplified Module Declaration
+
+## Problem
+
+Declaring a Magento module requires following files:
+
+* add `composer.json` - to declare package name, description, version, dependencies, etc
+* add `module.xml` - to declare module name and version dependencies
+* add `registration.php` - to let Magento know where the module resides, for Magento to be able to read configuration files
+
+The presense of 3 files causes duplication and complicated module management and sorting mechanisms. Current module sorting mechanism is inefficient.
+
+Every `registration.php` is included by Magento on every request. While with opcache this does not cause signfiicant performance impact, it can be avoided.
+
+## Solution
+
+* Rely on **composer** for module management, eliminate part of Magento module management behavior.
+* Register composer installation hook that will perform magento-specific module management.
+* Eliminate `registration.php` files.
+  * Composer installation hook will generate a 'magento_components.php' file that will contain all magento components and their installation paths. Only this file will be included on every request.
+  * Move module sorting behavior to composer installation hook. Use third-party topological sorting library. Write sorted modules to `magento_components.php`. This will move sorting behavior from run time to build time.
```

Comment:
Just a note, currently modules are not sorted in run time either, the sequence is calculated and written to `config.php` during deployment (and during modules installation/upgrade)

Reply:
Good point. Removed the mention of run-time improvement. Btw, using config.php instead of magento_components.php is worth exploring.